### PR TITLE
Fix beatmap removed post-creation in multiplayer tests

### PR DESCRIPTION
**Pull request by smoogipoo**
The main part of this is the addition of `GetOnlineBeatmapSet()`, which is overridden in tests to provide the given beatmap set.

### CODE REVIEW

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs [hunk 1]
```diff
@@ -89,13 +98,28 @@ public void ChangeUserBeatmapAvailability(int userId, BeatmapAvailability newBea
...
+            {
+                Settings =
+                {
+                    Name = apiRoom.Name.Value,
+                    BeatmapID = apiRoom.Playlist.Last().BeatmapID,
+                    RulesetID = apiRoom.Playlist.Last().RulesetID,
+                    BeatmapChecksum = apiRoom.Playlist.Last().Beatmap.Value.MD5Hash,
+                    RequiredMods = apiRoom.Playlist.Last().RequiredMods.Select(m => new APIMod(m)).ToArray(),
+                    AllowedMods = apiRoom.Playlist.Last().AllowedMods.Select(m => new APIMod(m)).ToArray(),
+                    PlaylistItemId = apiRoom.Playlist.Last().ID
```
Comment (bdach) [inline osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:118]: All multiplayer tests that join a room at start are failing on this (correctly) due to the playlist being empty. Honestly I'm not quite sure how it ever even worked in the first place. I think it didn't and the beatmap was shoved in at a later stage...?

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
For what it's worth, something like the following diff seems to resolve *most* failures:

```diff
diff --git a/osu.Game/Tests/Visual/Multiplayer/MultiplayerTestScene.cs b/osu.Game/Tests/Visual/Multiplayer/MultiplayerTestScene.cs
index 2e8c834c65..7775c2bd24 100644
--- a/osu.Game/Tests/Visual/Multiplayer/MultiplayerTestScene.cs
+++ b/osu.Game/Tests/Visual/Multiplayer/MultiplayerTestScene.cs
@@ -7,8 +7,10 @@
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Rooms;
 using osu.Game.Screens.OnlinePlay;
 using osu.Game.Screens.OnlinePlay.Lounge.Components;
+using osu.Game.Tests.Beatmaps;
 
 namespace osu.Game.Tests.Visual.Multiplayer
 {
@@ -48,7 +50,16 @@ protected MultiplayerTestScene(bool joinRoom = true)
             RoomManager.Schedule(() => RoomManager.PartRoom());
 
             if (joinRoom)
+            {
+                Room.Name.Value = "test name";
+                Room.Playlist.Add(new PlaylistItem
+                {
+                    Beatmap = { Value = new TestBeatmap(Ruleset.Value).BeatmapInfo },
+                    Ruleset = { Value = Ruleset.Value }
+                });
+
                 RoomManager.Schedule(() => RoomManager.CreateRoom(Room));
+            }
         });
 
         public override void SetUpSteps()
```

`TestSceneMultiplayerRoomManager` also needs similar logic. Also the name specification is mandatory as-is, otherwise `MultiplayerRoomSettings` dies on a nullref.

**Comment by peppy** [comment]
Is this aiming to fix a test failure somewhere?

**Comment by bdach** [comment]
The above patch is intended as a proposal on how to fix the test failures, yeah.

**Comment by bdach** [comment]
There are still remaining test failures that I'd indicated would be there (in `TestSceneMultiplayerRoomManager`) but I'm frankly too lost in the structure to be sure what's going wrong. From what I can tell even populating the new created rooms in that test doesn't help. `apiRoom` still ends up being without a playlist.

I'm not even entirely sure why for example `TestMultiplayerRoomJoinedWhenAPIRoomJoined` is named that exactly.

**Comment by peppy** [comment]
I meant the PR in general, rather than the diff. Wondering if this came up in some future use case or if it's the result of a CI failure somewhere. Lacking context.

**Comment by smoogipoo** [comment]
It was originally intended to improve the testability, but I'm not using it anymore. I noticed that when the match is created (e.g. in `TestSceneMultiplayerMatchSubScreen`), the beatmap is not displayed in the panel in the room, because it was being cleared via https://github.com/ppy/osu/blob/8a660b2bf545d886412be1b114f0c9041e661b06/osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs#L512-L514 and then the retrieve beatmap request was failing.

**Comment by bdach** [comment]
I think this change is nice to have. I did notice before that the beatmap was disappearing in the tests but chose to ignore it in favour of other things. Definitely good to have the multi tests behave the same way the real thing does in my book.

**Review by bdach — APPROVED** [review]

**Comment by smoogipoo** [comment]
For the most part you can go through an entire multi lobby in `TestSceneMultiplayer`, at least up until the results screen which you can't exit because there's no back button >_>

**Review by peppy — APPROVED** [review]
Looks logical enough

**Pull request merged**